### PR TITLE
react/week4

### PR DESCRIPTION
**Pull request by Rasim-Deniz**

### CODE REVIEW

File: react/week4/App.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import React from "react";
+import "./App.css";
+import StateProvider from "./context.js";
```
Comment (senner007) [inline react/week4/App.js]: See if you can come up with more descriptive names. State and Main are very generic names

File: react/week4/context.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+export const StateContext = createContext();
+
+const StateProvider = (props) => {
+  const [loading, setLoading] = useState(false);
+  const [users, setUsers] = useState([]);
+
+  const fetchUsers = (query) => {
+    setLoading(true);
+    fetch(`https://api.github.com/search/users?q=${query}`)
```
Comment (senner007) [inline react/week4/context.js]: I suggest using async with try/catch here. It is more readable, imo

File: react/week4/mainForm.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import React, { useContext } from "react";
+import { StateContext } from "./context.js";
+
+export default function MainForm() {
+  const { users, loading, fetchUsers } = useContext(StateContext);
```
Comment (senner007) [inline react/week4/mainForm.js]: Remember Single responsibility principle. This component has more than one responisbility. The display and input should be in different components

File: react/week4/mainForm.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            fetchUsers(e.target.value);
+          }}
+        />
+      </div>
+      <div>
+        {loading && <div>Loading... </div>}
+        {users.length === 0 || !users ? (
+          <h2>No results</h2>
+        ) : (
+          users.map((user) => <div key={user.id}> {user.login}</div>)
```
Comment (senner007) [inline react/week4/mainForm.js]: See if you can add some debouncing logic as well. You can reuse the logic we saw with the timer with the timeout being cleared in the cleanup function.

**Review by senner007 — APPROVED** [review]
Very good.

Just a few comments

File: react/week4/mainForm.js (hunk 4 above)
Comment (Rasim-Deniz) [reply react/week4/mainForm.js]: Thanks for your time and consideration. I guess, I didn't understand the concept of the debouncing logic well. May I get a bit more of a help from you about it?

File: react/week4/App.js (hunk 1 above)
Comment (Rasim-Deniz) [reply react/week4/App.js]: Are the recent names better now?

File: react/week4/context.js (hunk 2 above)
Comment (Rasim-Deniz) [reply react/week4/context.js]: I implemented it. Is it correct like this?

File: react/week4/mainForm.js (hunk 3 above)
Comment (Rasim-Deniz) [reply react/week4/mainForm.js]: Separated them, is it how it should be?

Comment (senner007) [reply react/week4/mainForm.js]: Or something along the lines of this maybe:

```
import React, { useContext, createContext, useState, useEffect } from "react";

function App() {
  return (
    <UsersSearch/>
  )
}

export const UserSearchContext = createContext();

function UsersSearch() {
  const [isLoading, setIsLoading] = useState(false);
  const [users, setUsers] = useState([]);
  const [userInput, setUserInput] = useState("");

  const handleErrors = (response) => {
    if (!response.ok) {
      throw Error(response.statusText);
    }
    return response;
  };

  useEffect(() => {
    (async () => {
      try {
        setIsLoading(true);
        const response = await fetch(
          `https://api.github.com/search/users?q=${userInput}`
        );
        handleErrors(response);
        const data = await response.json();
        const usersData = await data.items;
        setUsers(usersData);
      } catch (err) {
        console.log(err);
      } finally {
        setIsLoading(false);
      }
    })();
  }, [userInput])

 
  return (
    <UserSearchContext.Provider value={{ users, userInput, setUserInput }}>
      <SearchForm />
      { isLoading ? <Loading /> : <UsersDisplay /> }
    </UserSearchContext.Provider>
  );
}

function Loading() {
  return (
    <div>Loading... </div>
  )
}

function UsersDisplay() {
    const { users} = useContext(UserSearchContext);
  
    return (
      <div className="App">
        {users.length === 0 || !users ? (
          <h2>No results</h2>
        ) : (
          users.map((user) => <div key={user.id}> {user.login}</div>)
        )}
      </div>
    );
  }

function SearchForm() {
    const { userInput, setUserInput } = useContext(UserSearchContext);
  
    return (
      <div className="App">
        <h1>Github user searcher</h1>
        <div>
          <input
            placeholder="Search for user"
            value={userInput}
            onKeyUp={(e) => {
              setUserInput(e.target.value);
            }}
          />
        </div>
      </div>
    );
  }

export default App;
```

File: react/week4/mainForm.js (hunk 4 above)
Comment (senner007) [reply react/week4/mainForm.js]: The above with debouncing and error handling. This one is more advanced.

```
import React, { useContext, createContext, useState, useEffect } from 'react';

function App() {
  return <UsersSearch />;
}

export const UserSearchContext = createContext();

function UsersSearch() {
  const [userInput, setUserInput] = useState('');
  const [users, isCalling, isDebouncing] = useDebouncer(userInput, request);

  const handleErrors = (response) => {
    if (!response.ok) {
      throw Error(response.statusText);
    }
    return response;
  };

  async function request() {
    try {
      const response = await fetch(
        `https://api.github.com/search/users?q=${userInput}`
      );
      handleErrors(response);
      const data = await response.json();
      return await data.items;
    } catch (err) {
      return err;
    }
   
  }

  return (
    <UserSearchContext.Provider value={{ users, userInput, setUserInput }}>
      <SearchForm />
      {isCalling || isDebouncing ? <Loading /> : <UsersDisplay />}
    </UserSearchContext.Provider>
  );
}

function Loading() {
  return <div>Loading... </div>;
}

function UsersDisplay() {
  const { users } = useContext(UserSearchContext);

  if (users instanceof Error) {
    return (
      <h2>An error has occurred</h2>
    )
  }

  if (!users || users.length === 0 ) {
    return (
      <h2>No results</h2>
    )
  }


  return (
    <div className='App'>
        {users.map((user) => <div key={user.id}> {user.login}</div>)}
    </div>
  );
}

function SearchForm() {
  const { userInput, setUserInput } = useContext(UserSearchContext);

  return (
    <div className='App'>
      <h1>Github user searcher</h1>
      <div>
        <input
          placeholder='Search for user'
          value={userInput}
          onChange={(e) => {
            setUserInput(e.target.value);
          }}
        />
      </div>
    </div>
  );
}

function useDebouncer(inputText, request) {
  const [isCalling, setIsCalling] = useState(false);
  const [callQuery, setCallQuery] = useState('');
  const [response, setResponse] = useState('');
  const [isDebouncing, setIsDebouncing] = useState(false);

  useEffect(() => {
    if (isCalling) return;
    if (callQuery === inputText) return;
    setIsDebouncing(true);
    const timerId = setTimeout(async () => {
      setIsDebouncing(false);
      setIsCalling(true);
      setCallQuery(inputText);
      const response = inputText !== '' ? await request() : '';
      setResponse(response);
      setIsCalling(false);
    }, 500);
    return () => clearTimeout(timerId);
  }, [inputText, isCalling]);

  return [response, isCalling, isDebouncing];
}

export default App;

```

Comment (Rasim-Deniz) [reply react/week4/mainForm.js]: Thanks for your great support. I'm still missing some knowledge about debouncer.

**Pull request merged**